source/agora/common/TransactionData.d

```diff
@@ -0,0 +1,284 @@
+
+    /// The number of bytes of transaction data
+    public size_t length;
+
+    /// If the input data is larger than MAX_SIZE, this value is true, if not false.
+    public bool is_oversize;
+
+    /***************************************************************************
+
+        Create a TransactionData from binary data
+
+        Params:
+            bin  = Binary data to store in this `TransactionData`
+
+    ***************************************************************************/
+
+    public this (scope const ubyte[] bin) pure nothrow @nogc @safe
+    {
+        if (bin.length > MAX_SIZE)
+            this.is_oversize = true;
+        else
+            this.is_oversize = false;
+
+        size_t offset;
+        foreach (elem; bin)
+        {
+            this.data[offset++] = elem;
+            if (offset == MAX_SIZE)
+                break;
+        }
+        this.length = offset;
+    }
+
+    /***************************************************************************
+
+        Create a TransactionData from an hexadecimal string representation
+
+        Params:
+            hexstr = String representation of the binary data in base 16.
+                     The hexadecimal prefix (0x) is optional.
+                     Can be upper or lower case.
+
+    ***************************************************************************/
+
+    public this (scope const(char)[] hexstr) pure nothrow @nogc @safe
+    {
+        if (hexstr[0] == '0' && (hexstr[1] == 'x' || hexstr[1] == 'X'))
+            hexstr = hexstr[2 .. $];
+
+        if (hexstr.length / 2 > MAX_SIZE)
+            this.is_oversize = true;
+        else
+            this.is_oversize = false;
+
+        auto range = hexstr.byChar.map!(std.ascii.toLower!(char)).map!(parseHex);
+        size_t offset;
+        foreach (chunk; range.chunks(2))
+        {
+            (() @trusted {
+                this.data[offset++] = cast(ubyte)((chunk[0] << 4) + chunk[1]);
+            })();
+            if (offset == MAX_SIZE)
+                break;
+        }
+        this.length = offset;
+    }
+
+    /// Print the transaction data
+    public void toString (scope void delegate(const(char)[]) @safe sink) const @safe
+    {
+        static immutable hex_digits = `0123456789abcdef`;
+
+        sink("0x");
+        char[2] hex;
+        this.data[0 .. this.length].each!(
+            (elem)
+            {
+                hex[0] = hex_digits[elem >> 4];
+                hex[1] = hex_digits[(elem & 0b0000_1111)];
+                sink(hex);
+            });
+    }
+
+    /// Support for Vibe.d serialization to JSON
+    public string toString () const @safe
+    {
+        size_t idx;
+        char[] buffer;
+        buffer.length = this.length * 2 + 2;
+        scope sink = (const(char)[] v) {
+            buffer[idx .. idx + v.length] = v;
+            idx += v.length;
+        };
+        this.toString(sink);
+        return buffer.idup;
+    }
```

Comment:
A large portion of this file seems to be a copy of `BitBlob`. I don't understand why you didn't just use `BitBlob`, and add the `oversize` flag, instead of copying the code.

Reply:
Now I know there's a way to use BitBlob right away, thank you for letting me know.

Reply:
I tried to use BitBlob, but it's not working.
Because BitBlob only supports fixed sizes.

Reply:
Those are separate points: We do not want to have a fixed size, for the reasons mentioned, but if we did, this should have used `BitBlob` instead of copying all the code over. It's a basic case where [composition](https://en.wikipedia.org/wiki/Object_composition) could have been used.